### PR TITLE
Enable zoom by box selection

### PR DESCRIPTION
**Pull request by benoitblanc**
Hi,

This PR introduces a new `enableZoomByBoxSelection` option for `Zoom(In|Out)Plugin` to allow user to zoom by box selection. I needed to update `MapSelection` to make it work, I don't think I breaks something for other plugins that use this component but I'm not sure...

I tried to reproduce behavior for Zoom buttons in QGIS Desktop.

Here is a screencast with `enableZoomByBoxSelection: false` for `ZoomInPlugin` (actual behavior) and `enableZoomByBoxSelection: true` for `ZoomOutPlugin`:

QWCZoomByBox

Thanks for the review !

_This feature has been funded by Grand Lyon Métropole https://www.grandlyon.com/_

### CODE REVIEW

File: components/MapSelection.jsx [hunk 1]
```diff
@@ -112,7 +114,9 @@ class MapSelection extends React.Component {
         if (this.props.geomType === "DragBox") {
             this.drawInteraction = new ol.interaction.DragBox({
                 className: 'selection-drag-box',
-                condition: ol.events.condition.shiftKeyOnly
+                condition: () => {
+                    return ["ZoomIn", "ZoomOut"].includes(this.props.currentTask) || ol.events.condition.shiftKeyOnly;
+                }
```
Comment (manisandro) [inline components/MapSelection.jsx:119]: I think you can generally simplify the condition to `return true`, `shiftKeyOnly` looks like leftover code, and I don't think it makes any sense here

File: plugins/ZoomButtons.jsx [hunk 2]
```diff
@@ -26,17 +29,58 @@ import './style/Buttons.css';
...
+            if (this.props.direction > 0) {
+                this.setState({disabled: this.props.currentZoom >= this.props.maxZoom, task: "ZoomIn"});
+            } else if (this.props.direction < 0) {
+                this.setState({disabled: this.props.currentZoom <= 0, task: "ZoomOut"});
+            }
+        }
+        if (this.props.currentTask !== null && this.props.currentTask === this.state.task && this.state.disabled) {
+            this.props.setCurrentTask(null);
+        }
+        if (this.props.currentTask === this.state.task && prevProps !== this.props) {
```
Comment (manisandro) [inline plugins/ZoomButtons.jsx:71]: If you write `this.props.click && this.props.click !== prevProps.click` here and directly use  `this.props.click.coordinate` below you can simplify the code

**Comment by manisandro** [comment]
A general remark: Does this conflict/overlap with the already implemented shift+drag zoomin?

File: components/MapSelection.jsx (hunk 1 above)
Comment (benoitblanc) [reply components/MapSelection.jsx]: I thought it was used by the shift+drag zoom already implemented but nope. I have removed all the condition

File: plugins/ZoomButtons.jsx (hunk 2 above)
Comment (benoitblanc) [reply plugins/ZoomButtons.jsx]: Resolved

File: components/MapSelection.jsx [hunk 3]
```diff
@@ -21,6 +21,7 @@ class MapSelection extends React.Component {
     static propTypes = {
         /* Whether the selection tool is active */
         active: PropTypes.bool,
+        currentTask: PropTypes.string,
```
Comment (manisandro) [inline components/MapSelection.jsx:24]: This can also be removed (as well as below)

File: plugins/ZoomButtons.jsx [hunk 4]
```diff
@@ -26,18 +29,53 @@ import './style/Buttons.css';
...
+            if (this.props.direction > 0) {
+                this.setState({disabled: this.props.currentZoom >= this.props.maxZoom, task: "ZoomIn"});
+            } else if (this.props.direction < 0) {
+                this.setState({disabled: this.props.currentZoom <= 0, task: "ZoomOut"});
+            }
+        }
+        if (this.props.currentTask !== null && this.props.currentTask === this.state.task && this.state.disabled) {
+            this.props.setCurrentTask(null);
+        }
+        if (this.props.currentTask === this.state.task && prevProps !== this.props && this.props.click !== prevProps.click) {
```
Comment (manisandro) [inline plugins/ZoomButtons.jsx:71]: > this.props.currentTask === this.state.task && this.props.click && this.props.click !== prevProps.click

Comment (benoitblanc) [reply plugins/ZoomButtons.jsx]: Thanks :facepalm:

**Comment by benoitblanc** [comment]
> I'd propose these additional changes [1], what do you think?
> 
> [1] [db228da](https://github.com/qgis/qwc2/commit/db228daa1386370ad6885b437467b329c94b46d6)

Yes I agree with all of your changes, thanks.

For `don't automatically exit task when zoom limit reached`, I added it at the end of my PR because I wanted to exit the task to not have a `zoom-in|out` cursor for the user when he could not zoom because he has reached the limit, but I'm ok to keep this as current task. So user could still draw drag-box but zoom should not have effect I guess

**Comment by manisandro** [comment]
>  So user could still draw drag-box but zoom should not have effect I guess

It will just re-center the map without changing the zoom level, which I think is not a bad behaviour.

**Comment by manisandro** [comment]
Manually merged, thanks!

**Pull request closed without merging**